lxd/daemon.go

```diff
@@ -1129,7 +1129,11 @@ func (d *Daemon) Ready() error {
 		d.tasks.Add(pruneExpireCustomVolumeSnapshotsTask(d))
 
 		// Take snapshot of custom volumes (minutely check of configurable cron expression)
-		d.tasks.Add(autoCreateCustomVolumeSnapshotsTask(d))
+		_, err := d.gateway.CurrentRaftNodes()
+		if err == nil {
+			// We're the leader
+			d.tasks.Add(autoCreateCustomVolumeSnapshotsTask(d))
```

Comment:
You can't do that. The leader is something which changes over time so your code may cause the task to start on multiple nodes or on none at all.

Instead the task itself should be run on all nodes but any given volume should only be snapshotted by a single node.
Ideally we also want to spread that load across the cluster and not have a single node handle them all.

I'd recommend you:
 - Run the task on all nodes
 - Determine whether to run a given snapshot or not based on a stable random for a given volume.

That last step is effectively done by:
 - Fetch a slice of all online cluster nodes (offline nodes can't snapshot, obviously)
 - Initialize a random number generator using volume.id as seed
 - Get random int in the online cluster node slice
 - If the node that is picked is the local one, perform the snapshot

This should ensure that so long as nodes aren't offline, a given volume always gets snapshotted by the same node.
If that node goes offline, another will handle it. It also spreads the snapshots across the cluster.

@tomponline can help you there as he's been doing that kind of stable random logic in our network code lately.

Reply:
@monstermunchkin yes, this section in the bridge network driver deals with generating a stable MAC address from the LXD server's certificate fingerprint, node ID (optional) and network ID: using the fnv hash algorithm: https://github.com/lxc/lxd/blob/master/lxd/network/driver_bridge.go#L534-L571

Let me know if you have any questions.